Storage/StorageManager.php

```diff
@@ -161,68 +118,177 @@ public function storeUploadedFile($type, UploadedFile $uploadedFile, $newFilenam
+
+    /**
+     * @param StorageFile $file
+     * @param UploadedFile|StreamInterface|string $content
+     * @param array $options
+     */
+    private function storeFileContent(StorageFile $file, $content, $options = [])
+    {
+        $storage = $file->getStorage();
+
+        $metadata = empty($options['metadata']) ? [] : $options['metadata'];
+
+        if (isset($options['realPath'])) {
+            $path = $options['realPath'];
+            $bytes = $this->streamCopy($path, $this->getIOStream($storage->getSlug(), $file->getKey()));
+
+            $file->setContentHash(md5_file($path));
+            $file->setSize($bytes);
+
+            if (!isset($metadata['ContentType'])) {
+                $metadata['ContentType'] = MimeTypeGuesser::getInstance()->guess($path);
+            }
+
+            $file->setFileMetadata($metadata);
+        } else {
+            if (!isset($metadata['ContentType'])) {
+                $metadata['ContentType'] = 'binary/octet-stream';
+            }
+
+            $bytes = $file->setContent($content, $metadata);
+        }
+
+        $file->setMimeType($metadata['ContentType']);
 
         if ($bytes == 0) {
             throw new \RuntimeException(
                 sprintf(
-                    "Unable to copy file to storage '%s' with key '%s' from local path '%s'.  Zero bytes copied.",
+                    "Unable to stream file to storage '%s' with key '%s'.  Zero bytes streamed.",
                     $storage->getName(),
-                    $file->getKey(),
-                    $path
+                    $file->getKey()
                 )
             );
         }
 
         if (!$file->exists()) {
             throw new \RuntimeException(
                 sprintf(
-                    "Could not stream copy to storage '%s' with key '%s'.  Resultant file does not exist.",
+                    "Could not stream to storage '%s' with key '%s'.  Resultant file does not exist.",
                     $storage->getName(),
                     $file->getKey()
                 )
             );
         }
 
-        // Ensure other fields on the file are set properly
-        $file->setMimeType($mimeType);
-        $file->setContentHash(md5_file($path));
-        $file->setSize($bytes);
-        $file->setType($type);
-        // ContentType is used by the AmazonAWSS3 Adapter
-        $file->setFileMetadata(['contentType' => $mimeType]);
+        $this->unlinkIfNeeded($options);
+    }
+
+    /**
+     * Return generated filename
+     *
+     * @param array $options
+     * @return string
+     */
+    private function getFilename($options)
+    {
+        if (isset($options['newFilename'])) {
+            return $options['newFilename'];
+        }
 
-        if ($unlinkAfterStore) {
-            unlink($path);
+        if (isset($options['originalFilename'])) {
+            $path = $options['originalFilename'];
+        } else if (isset($options['realPath'])) {
+            $path = $options['realPath'];
```

Comment:
These options appear to be doing the same thing??

Reply:
Yes, but the source is received from different places. for an uploaded file we would have realPath something like /tmp/1123114 without extension, so we should use originalFilename